validator/sawtooth_validator/journal/publisher.py

```diff
@@ -370,7 +370,8 @@ def _finalize_block(self, block):
                 else:
                     block.add_batch(batch)
                     self._committed_txn_cache.add_batch(batch)
-                state_hash = result.state_hash
+                if result.state_hash is not None:
+                    state_hash = result.state_hash
```

Comment:
It is more about how data is being passed back to users of the scheduler. No single txn will get a new state hash, and no single batch will get a new state hash, but the result of all of the valid batches will get a state hash (unless it is happening on block validation). So, At most one of the result.state_hash will be not None, and only that one should be recorded in the variable state_hash.